### ISSUE TITLE
GraphQL Query running out of memory when trying to fetch lots of files

### ISSUE THREAD

**Issue by maxime-rainville** [silverstripe/silverstripe-asset-admin#828]
## Pull Requests
- [x] https://github.com/silverstripe/silverstripe-asset-admin/pull/886

Our super secret project has a folder with 45K files. The GraphQL request to retrieve the list of files to display run for 15 sec before running out of memory.

That's the body of the request.
```
{
  "operationName":"ReadFiles",
  "variables":
  {
    "rootFilter":
      {
        "id":1,
        "anyChildId":null
      },
    "childrenFilter":{"recursive":false},
    "limit":50,
    "offset":0
  },
  "query":"
query ReadFiles($limit: Int!, $offset: Int!, $rootFilter: FileFilterInput, $childrenFilter: FileFilterInput, $sortBy: [ChildrenSortInputType]) {
  readFiles(filter: $rootFilter) {
    pageInfo {
      totalCount
      __typename
    }
    edges {
      node {
        ...FileInterfaceFields
        ...FileFields
        ... on Folder {
          children(limit: $limit, offset: $offset, filter: $childrenFilter, sortBy: $sortBy) {
            pageInfo {
              totalCount
              __typename
            }
            edges {
              node {
                ...FileInterfaceFields
                ...FileFields
                ...FolderFields
                __typename
              }
              __typename
            }
            __typename
          }
          parents {
            id
            title
            __typename
          }
          __typename
        }
        __typename
      }
      __typename
    }
    __typename
  }
}

fragment FileInterfaceFields on FileInterface {
  canDelete
  canEdit
  canView
  category
  exists
  filename
  id
  lastEdited
  name
  parentId
  title
  type
  url
  __typename
}

fragment FileFields on File {
  draft
  extension
  published
  modified
  size
  smallThumbnail
  thumbnail
  inUseCount
  __typename
}

fragment FolderFields on Folder {
  filesInUseCount
  __typename
}
  "
}
```

**Comment by robbieaverill** [comment]
Fixed with #8495

**Comment by micmania1** [comment]
This is still a problem when used with 45,000 real images. The fix was for purely for the number of database records.

**Comment by chillu** [comment]
I'm hoping that we don't need 45k files in order to reproduce this, but can rather monitor memory usage on files through blackfire.io, and extrapolate from that. I've got a few quickfire theories:

- Thumbnail generation craps out on number of files due to memory leaks in PHP or GD
- Thumbnail generation craps out on a particularly large image (not the amount of images)
- ORM loads too many records, not paginating effectively

**Comment by chillu** [comment]
I've talked to @unclecheese before. It sounds like we're querying the whole data set, and running it through canView in memory. Which will hit either CPU or memory limits. The thing is that it's been a known issue in framework for many years in other contexts (GridField, and ComplexTableField before that) - see https://github.com/silverstripe/silverstripe-framework/issues/5995. It can only be fixed by not allowing procedural definition of canView (and using declarative systems like an ACL), which is a major change to how developers use SilverStripe.

So in conclusion, I don't see how we need to spend effort fixing this for a list in AssetAdmin produced through GraphQL when it's a long standing issue for virtually every other list in the system (all GridFields).

There's UX improvements we can do there (show blanked out rows rather than filter them out), but that's a separate issue, and lower priority. Happy for someone to write up.

Note: The project which raised this isn't even using canView for files, but just wants the performance issues arising from this fixed.

**Comment by unclecheese** [comment]
Fix is posted to the top of the ticket.
